baybe/searchspace/continuous.py

```diff
@@ -186,61 +228,153 @@ def transform(
         import torch
         from botorch.utils.sampling import get_polytope_samples
 
-        # TODO Revisit: torch and botorch here are actually only necessary if there
-        # are constraints. If there are none and the lists are empty we can just sample
-        # without the get_polytope_samples, which means torch and botorch
-        # wouldn't be needed.
-
         points = get_polytope_samples(
-            n=n_points,
-            bounds=torch.from_numpy(self.param_bounds_comp),
+            n=batch_size,
+            bounds=torch.from_numpy(bounds),
             equality_constraints=[
                 c.to_botorch(self.parameters) for c in self.constraints_lin_eq
             ],
             inequality_constraints=[
                 c.to_botorch(self.parameters) for c in self.constraints_lin_ineq
             ],
         )
-
         return pd.DataFrame(points, columns=self.param_names)
 
+    def _sample_with_cardinality_constraints(self, batch_size: int) -> pd.DataFrame:
+        """Draw random samples from a polytope with cardinality constraints."""
+        if not self.constraints_cardinality:
+            raise RuntimeError(
+                f"This method should not be called without any constraints of type "
+                f"'{ContinuousCardinalityConstraint.__name__}' in place. "
+                f"Use '{SubspaceContinuous._sample_from_bounds.__name__}' "
+                f"or '{SubspaceContinuous._sample_from_polytope.__name__}' instead."
+            )
+
+        # List to store the created samples
+        samples: list[pd.DataFrame] = []
+
+        # Counter for failed sampling attempts
+        n_fails = 0
+
+        while len(samples) < batch_size:
+            # Randomly set some parameters inactive
+            inactive_params_sample = self._sample_inactive_parameters(1)[0]
+
+            # Remove the inactive parameters from the search space
+            subspace_without_cardinality_constraint = self._drop_parameters(
+                inactive_params_sample
+            )
+
+            # Sample from the reduced space
+            try:
+                sample = subspace_without_cardinality_constraint.sample(1)
+                samples.append(sample)
+            except Exception:
+                n_fails += 1
+
+            # Avoid infinite loop
+            if n_fails >= _MAX_CARDINALITY_SAMPLING_ATTEMPTS:
+                raise RuntimeError(
+                    f"The number of failed sampling attempts has exceeded the limit "
+                    f"of {_MAX_CARDINALITY_SAMPLING_ATTEMPTS}. "
+                    f"It appears that the feasible region of the search space is very "
+                    f"small. Please review the search space constraints."
+                )
+
+        # Combine the samples and fill in inactive parameters
+        parameter_names = [p.name for p in self.parameters]
+        return pd.concat(samples).reindex(columns=parameter_names).fillna(0.0)
```

Comment:
shouldnt this be `DTypeNumpyFloat(0.0)` from `utils.numeric` ?

Reply:
as discussed, probably "no" (but let's see how long it takes to haunt us 😄 )